test-data/unit/check-narrowing.test

```diff
@@ -0,0 +1,406 @@
+    reveal_type(impossible_mixture)     # N: Revealed type is 'Union[__main__.KeyedObject, TypedDict('__main__.KeyedTypedDict', {'key': Literal[__main__.Key.B]})]'
+else:
+    reveal_type(impossible_mixture)     # N: Revealed type is 'Union[__main__.KeyedObject, TypedDict('__main__.KeyedTypedDict', {'key': Literal[__main__.Key.B]})]'
+
+weird_mixture: Union[KeyedTypedDict, KeyedNamedTuple]
+if weird_mixture["key"] is Key.B:       # E: Invalid tuple index type (actual type "str", expected type "Union[int, slice]")
+    reveal_type(weird_mixture)          # N: Revealed type is 'Union[TypedDict('__main__.KeyedTypedDict', {'key': Literal[__main__.Key.B]}), Tuple[Literal[__main__.Key.C], fallback=__main__.KeyedNamedTuple]]'
+else:
+    reveal_type(weird_mixture)          # N: Revealed type is 'Union[TypedDict('__main__.KeyedTypedDict', {'key': Literal[__main__.Key.B]}), Tuple[Literal[__main__.Key.C], fallback=__main__.KeyedNamedTuple]]'
+
+if weird_mixture[0] is Key.B:           # E: TypedDict key must be a string literal; expected one of ('key')
+    reveal_type(weird_mixture)          # N: Revealed type is 'Union[TypedDict('__main__.KeyedTypedDict', {'key': Literal[__main__.Key.B]}), Tuple[Literal[__main__.Key.C], fallback=__main__.KeyedNamedTuple]]'
+else:
+    reveal_type(weird_mixture)          # N: Revealed type is 'Union[TypedDict('__main__.KeyedTypedDict', {'key': Literal[__main__.Key.B]}), Tuple[Literal[__main__.Key.C], fallback=__main__.KeyedNamedTuple]]'
+[builtins fixtures/slice.pyi]
+
+[case testNarrowingParentWithProperties]
+from enum import Enum
+from typing import Union
+from typing_extensions import Literal
+
+class Key(Enum):
+    A = 1
+    B = 2
+    C = 3
+
+class Object1:
+    key: Literal[Key.A]
+
+class Object2:
+    @property
+    def key(self) -> Literal[Key.A]: ...
+
+class Object3:
+    @property
+    def key(self) -> Literal[Key.B]: ...
+
+x: Union[Object1, Object2, Object3]
+if x.key is Key.A:
+    reveal_type(x)  # N: Revealed type is 'Union[__main__.Object1, __main__.Object2]'
+else:
+    reveal_type(x)  # N: Revealed type is '__main__.Object3'
+[builtins fixtures/property.pyi]
+
+[case testNarrowingParentWithAny]
+from enum import Enum
+from typing import Union, Any
+from typing_extensions import Literal
+
+class Key(Enum):
+    A = 1
+    B = 2
+    C = 3
+
+class Object1:
+    key: Literal[Key.A]
+
+class Object2:
+    key: Literal[Key.B]
+
+x: Union[Object1, Object2, Any]
+if x.key is Key.A:
+    reveal_type(x.key)  # N: Revealed type is 'Literal[__main__.Key.A]'
+    reveal_type(x)      # N: Revealed type is 'Union[__main__.Object1, Any]'
+else:
+    # TODO: Is this a bug? Should we skip inferring Any for singleton types?
+    reveal_type(x.key)  # N: Revealed type is 'Union[Any, Literal[__main__.Key.B]]'
+    reveal_type(x)      # N: Revealed type is 'Union[__main__.Object1, __main__.Object2, Any]'
+
+[case testNarrowingParentsHierarchy]
```

Comment:
Just in case I would add a test involving "double index" like `if x['model']['kind'] is ...: ...` for nested typed dicts.